### PR TITLE
test: Comparability and hashability of `PackageInfo` and `MatchSpec` 

### PR DESCRIPTION
**Pull request by jjerphan**
Follow-up of #3363.

This adds just some simply test for `PackageInfo` and `MatchSpec` for now.

### CODE REVIEW

File: libmamba/include/mamba/specs/regex_spec.hpp [hunk 1]
```diff
@@ -75,4 +75,13 @@ struct fmt::formatter<mamba::specs::RegexSpec>
...
 
+template <>
+struct std::hash<mamba::specs::RegexSpec>
+{
+    auto operator()(const mamba::specs::RegexSpec& spec) const -> std::size_t
+    {
+        return std::hash<std::string>{}(spec.str());
+    }
+};
+
```
Comment (Hind-M) [inline libmamba/include/mamba/specs/regex_spec.hpp:86]: I didn't check thoroughly all what have been changed with #3363 compared to before, so was this something missing from #3363?

File: libmamba/include/mamba/util/flat_set.hpp [hunk 2]
```diff
@@ -565,7 +565,7 @@ struct std::hash<mamba::util::flat_set<Key, Compare, Allocator>>
 {
     auto operator()(const mamba::util::flat_set<Key, Compare, Allocator>& set) const -> std::size_t
     {
-        return mamba::util::hash_vals(set);
+        return mamba::util::hash_range(set);
```
Comment (Hind-M) [inline libmamba/include/mamba/util/flat_set.hpp:568]: Why this change?

File: libmamba/include/mamba/util/heap_optional.hpp [hunk 3]
```diff
@@ -251,17 +251,4 @@ namespace mamba::util
...
-    std::size_t operator()(const mamba::util::heap_optional<T>& opt) const
-    {
-        if (opt.has_value())
-        {
-            return std::hash<T>{}(*opt);
-        }
-        return 0;
-    }
-};
-
```
Comment (Hind-M) [inline libmamba/include/mamba/util/heap_optional.hpp:266]: This is not used anymore?

File: libmamba/include/mamba/specs/regex_spec.hpp (hunk 1 above)
Comment (jjerphan) [reply libmamba/include/mamba/specs/regex_spec.hpp:86]: This was missing from #3363, yes.

File: libmamba/include/mamba/util/flat_set.hpp (hunk 2 above)
Comment (jjerphan) [reply libmamba/include/mamba/util/flat_set.hpp:568]: `hash_vals` hashes `set` as a value rather than its' internal value (creating an infinite recursion).

`hash_range` hashes elements considers `set` a range and hashes all of its elements.

File: libmamba/include/mamba/util/heap_optional.hpp (hunk 3 above)
Comment (jjerphan) [reply libmamba/include/mamba/util/heap_optional.hpp:266]: Yes, this is not needed thanks to [this line](https://github.com/mamba-org/mamba/pull/3369/files#diff-7b3f6d24884e8d3e7e3a160ab570bdfe231bf8a13df239de5818e98581840ab5R1049) which hashes the value contained by the `heap_optional`.

IIRC, @Klaim has it in mind and I understood why.

File: libmamba/include/mamba/util/flat_set.hpp (hunk 2 above)
Comment (Hind-M) [reply libmamba/include/mamba/util/flat_set.hpp:568]: I'm actually wondering about the use cases of using `hash_vals`, and btw I just realized there are no tests for `hash_vals`. Would you mind adding them please? (in this PR or another one, as you prefer).

**Review by Hind-M — APPROVED** [review]
Can you please rebase now that #3371 is merged? Thanks.

File: libmamba/include/mamba/util/flat_set.hpp (hunk 2 above)
Comment (jjerphan) [reply libmamba/include/mamba/util/flat_set.hpp:568]: `hash_vals` is used to combine the hashes of individual scalar values. I can open a PR after this one to test its behavior.
